capsules/src/adc.rs

```diff
@@ -1,77 +1,727 @@
+    /// ownership of buffers to the capsule.
+    fn stop_sampling(&self) -> ReturnCode {
+        if !self.active.get() || self.mode.get() == AdcMode::NoMode {
+            // already inactive!
+            return ReturnCode::SUCCESS;
+        }
+
+        // we're going to leave active as true here, but set ourselves to NoMode so the client
+        // handler functions will clean up state properly when the ADC is ready again
+        self.mode.set(AdcMode::NoMode);
+        self.app_buf_offset.set(0);
+
+        // actually cancel the operation
+        self.adc.stop_sampling()
     }
 }
 
-impl<'a, A: AdcSingle + 'a> Client for ADC<'a, A> {
-    fn sample_done(&self, sample: u16) {
-        self.channel.get().map(|cur_channel| {
-            self.channel.set(None);
-            self.app.each(|app| if app.channel == Some(cur_channel) {
-                app.channel = None;
-                app.callback.map(|mut cb| cb.schedule(0, cur_channel as usize, sample as usize));
-            } else if app.channel.is_some() {
-                self.channel.set(app.channel);
+/// Callbacks from the ADC driver
+impl<'a, A: hil::adc::Adc + hil::adc::AdcHighSpeed + 'a> hil::adc::Client for Adc<'a, A> {
+    /// Single sample operation complete
+    /// Collects the sample and provides a callback to the application
+    ///
+    /// sample - analog sample value
+    fn sample_ready(&self, sample: u16) {
+        if self.active.get() && self.mode.get() == AdcMode::SingleSample {
+            // single sample complete, clean up state
+            self.active.set(false);
+            self.mode.set(AdcMode::NoMode);
+
+            // perform callback
+            self.callback.get().map(|mut callback| {
+                callback.schedule(AdcMode::SingleSample as usize,
+                                  self.channel.get(),
+                                  sample as usize);
             });
-        });
-        self.channel.get().map(|next_channel| { self.adc.sample(next_channel); });
+
+        } else if self.active.get() && self.mode.get() == AdcMode::ContinuousSample {
+            // sample ready in continuous sampling operation, keep state
+
+            // perform callback
+            self.callback.get().map(|mut callback| {
+                callback.schedule(AdcMode::ContinuousSample as usize,
+                                  self.channel.get(),
+                                  sample as usize);
+            });
+
+        } else {
+            // operation probably canceled. Make sure state is consistent. No callback
+            self.active.set(false);
+            self.mode.set(AdcMode::NoMode);
+        }
     }
 }
 
-impl<'a, A: AdcSingle + 'a> Driver for ADC<'a, A> {
+/// Callbacks from the High Speed ADC driver
+impl<'a, A: hil::adc::Adc + hil::adc::AdcHighSpeed + 'a> hil::adc::HighSpeedClient for Adc<'a, A> {
+    /// Internal buffer has filled from a buffered sampling operation.
```

Comment:
Can you reduce all of the comments to 80 columns, so they read easily in an 80 column display? The code is almost entirely within 80 columns, there are just a few lines that sneak over, they're fine.